File: python/src/roax_canon/flatten.py
"""Flattening a record into typed leaves (specification section 3.3).

::

    flatten(node, path):
      if node is a map and is empty:        emit (path, typeTag(path, "object"), -)
      if node is a map and is non-empty:    for each key k: flatten(node[k], path ‖ KEY(k))
      if node is an array and is empty:     emit (path, typeTag(path, "array"), -)
      if node is an array and is non-empty: for each index i: flatten(node[i], path ‖ INDEX(i))
      otherwise:                            emit (path, typeTag(path, jsonKind(node)), node)

Map iteration order is irrelevant, because leaves are sorted by encoded path in
specification section 9.
That is deliberate: it is what lets a Go implementation work despite Go randomizing map
iteration order, and it removes the class of bug OpenAttestation inherits from JavaScript
key enumeration.

An empty container is a leaf so that removing it changes the root.
Empty array, empty object and explicit null are **three distinct leaves**, which is a
departure from dogtag, where `crates/dogtag-standard-rs/src/flatten.rs:96-115` collapses
all three to ``TypedScalar::Null``.
"""

from __future__ import annotations

from typing import Any, Sequence

from .errors import ErrorCode, InputError, RoaxError
from .jsonio import json_kind
from .leaf import Leaf
from .path import Index, Key, Segment, display_path
from .text import nfc
from .typemap import TypeResolver
from .value import (
    BLOB_REF,
    BYTES,
    EMPTY_ARRAY,
    EMPTY_OBJECT,
    TAG_NAMES,
    decode_base64_canonical,
)

__all__ = ["RESERVED_KEY_PREFIX", "check_reserved_namespace", "flatten"]

#: Specification section 11.2: the reserved ASCII prefix, including the dot.
#: A key named ``roax`` or ``roaxX``, with no dot, collides with nothing and is accepted.
RESERVED_KEY_PREFIX = "roax."


def check_reserved_namespace(segments: Sequence[Segment]) -> None:
    """The reserved-namespace guard (specification section 11.2).

    > **Normative:** no record-supplied path may have, as its **first segment**, a `KEY`
    > whose NFC-normalized key begins with the ASCII prefix ``roax.``.

    Three ways to get this wrong, all closed here:

    * it is not a test against a rendered display path, and this function never builds one
      (specification section 5.2);
    * it applies to the **first** segment only, because every reserved path is a single
      segment, so ``[KEY("a"), KEY("roax.foo")]`` differs from every reserved path in
      segment count and cannot collide;
    * it compares the **NFC-normalized** key, not the bytes as received, because section
      6.1 normalizes keys before they are hashed and a check on the raw bytes is checking
      a different string from the one that gets committed.

    The specification states honestly that the third is currently unobservable, since no
    character normalizes into ``roax.``.
    It is implemented anyway, for the reason it gives: this guard and the hashing path
    must reason about the same string.
    """
    if not segments:
        return
    first = segments[0]
    if not isinstance(first, Key):
        return
    if nfc(first.value, where="first path segment").startswith(RESERVED_KEY_PREFIX):
        raise InputError(
            ErrorCode.RESERVED_NAMESPACE,
            f"record-supplied first segment {first.value!r} is in the reserved "
            f"{RESERVED_KEY_PREFIX!r} namespace (specification section 11.2)",
        )


def _coerce_record_value(tag: int, node: Any, segments: Sequence[Segment]) -> Any:
    """Turn a record scalar into the carrier :func:`encode_value` expects.

    The only conversion is base64: a field bound as `BYTES` carries base64 text in the
    record and the encoded value is the decoded bytes (specification section 6.3), which
    is why the canonical base64 form is pinned there rather than left to be discovered.
    A field bound as `STRING` over that same text is hashed as the text, so no base64 rule
    enters its digest.
    """
    if tag == BYTES:
        if isinstance(node, (bytes, bytearray)):
            return bytes(node)
        if isinstance(node, str):
            return decode_base64_canonical(node)
        raise RoaxError(
            ErrorCode.ENVELOPE_SHAPE,
            f"{display_path(segments)}: BYTES binding needs base64 text or bytes",
        )
    if tag == BLOB_REF:
        raise RoaxError(
            ErrorCode.BLOB_REF_NOT_DECLARED,
            f"{display_path(segments)}: no version-1 profile declares tag 8 BLOB_REF "
            f"(specification section 6.5)",
        )
    return node


def flatten(
    record: Any,
    resolver: TypeResolver,
    *,
    authorize_empty_containers: bool = True,
) -> list[Leaf]:
    """Flatten a record into typed leaves, failing closed on any unresolved path.

    ``authorize_empty_containers`` selects between two readings of specification section
    3.3, and the default is the one the specification text states.

    * ``True``: an empty container's tag is taken from the type map, so `EMPTY_ARRAY` and
      `EMPTY_OBJECT` are emitted "only when the exact selected map authorizes that
      structured path and observed kind under section 4.2", and an unauthorized empty
      container fails closed.
      Section 3.3 gives the reason: assigning tags 6 or 7 before map resolution would let
      an unknown empty issuer extension bypass decision D7's fail-closed rule.
    * ``False``: an empty container is tagged structurally, without consulting the map.

    **The committed conformance corpus requires ``False`` and this is a real divergence,
    reported rather than smoothed over.**
    `corpus/type-maps/org.roax.corpus.synthetic.json` binds ``a.b`` only at kind ``null``
    and ``a.**`` only at kind ``string``, yet `corpus/fixtures/records/structure-empty-
    array.json` and `structure-empty-object.json` carry ``a.b`` as ``[]`` and ``{}`` and
    the class-5 vectors assert a root for each.
    Under the section 3.3 rule those two records fail closed and have no root at all.
    The corpus predates the type-map artifact work that added that sentence, so the
    divergence is one of vintage rather than of reading, and specification section 1.1
    makes the specification govern.
    :mod:`roax_canon.corpus` passes ``False`` with that citation attached; every other
    caller gets the specification's rule.
    """
    leaves: list[Leaf] = []

    def walk(node: Any, segments: tuple[Segment, ...]) -> None:
        kind = json_kind(node)

        if kind == "object":
            if not node:
                leaves.append(Leaf(segments, _container_tag(segments, "object"), None))
                return
            for key, child in node.items():
                child_path = segments + (Key(key),)
                if len(child_path) == 1:
                    check_reserved_namespace(child_path)
                walk(child, child_path)
            return

        if kind == "array":
            if not node:
                leaves.append(Leaf(segments, _container_tag(segments, "array"), None))
                return
            for i, child in enumerate(node):
                walk(child, segments + (Index(i),))
            return

        tag = resolver.resolve(segments, kind)
        if tag in (EMPTY_ARRAY, EMPTY_OBJECT):
            raise RoaxError(
                ErrorCode.TYPE_MAP_REJECTED,
                f"{display_path(segments)}: map binds a scalar to "
                f"{TAG_NAMES[tag]}, which only a container may carry",
            )
        leaves.append(Leaf(segments, tag, _coerce_record_value(tag, node, segments)))

    def _container_tag(segments: tuple[Segment, ...], kind: str) -> int:
        structural = EMPTY_OBJECT if kind == "object" else EMPTY_ARRAY
        if not authorize_empty_containers:
            return structural
        tag = resolver.resolve(segments, kind)
        if tag != structural:
            raise RoaxError(
                ErrorCode.TYPE_MAP_REJECTED,
                f"{display_path(segments)}: map binds an empty {kind} to "
                f"{TAG_NAMES.get(tag, tag)} rather than {TAG_NAMES[structural]}",
            )
        return structural

    walk(record, ())

    if not leaves:
        raise InputError(
            ErrorCode.EMPTY_RECORD,
            "a record that contributes zero leaves of its own MUST be rejected at "
            "issuance (specification section 3.3)",
        )
    return leaves

File: kotlin/roax-canon/src/main/kotlin/io/roax/canon/Flatten.kt
package io.roax.canon

import io.roax.canon.json.JsonArray
import io.roax.canon.json.JsonBoolean
import io.roax.canon.json.JsonNull
import io.roax.canon.json.JsonNumber
import io.roax.canon.json.JsonObject
import io.roax.canon.json.JsonString
import io.roax.canon.json.JsonValue

/** One `(path, typeTag, value)` triple, before a salt is attached. */
data class FlatLeaf(val segments: List<Segment>, val tag: TypeTag, val value: RoaxValue)

/**
 * How the flattener treats an empty container whose `(path, observed kind)` the selected map does
 * not bind.
 *
 * **This is a switch because the specification and the committed corpus disagree, and
 * `docs/typescript-implementation-findings.md` asks an implementer to expose both readings rather
 * than pick one silently.**
 *
 * ROAX-CANON/1 section 3.3 is unambiguous: an empty container gets tag 6 or 7 "only when the exact
 * selected map authorizes that structured path and observed kind under section 4.2", because
 * assigning the tag first would let an unknown empty issuer extension bypass decision D7's
 * fail-closed rule. But `corpus/type-maps/org.roax.corpus.synthetic.json` declares `a.b` for
 * `jsonKind: "null"` alone, while two class-5 fixtures carry an empty array and an empty object
 * there - so under the specification's own rule both records fail closed and have no root, and the
 * corpus asserts one.
 *
 * Under section 1.1 that is a release-blocking corpus defect, and the narrow fix is a corpus edit
 * rather than an implementation default.
 */
enum class EmptyContainerAuthorization {

    /** ROAX-CANON/1 section 3.3 as written. The default, because the specification governs. */
    REQUIRED,

    /**
     * The reading the committed corpus was built under: an empty container is tagged 6 or 7 from
     * its observed kind without consulting the map.
     *
     * Selecting this makes those two class-5 vectors pass and **weakens decision D7 exactly where
     * section 3.3 warns**, so it is named for the artifact it exists to reproduce rather than
     * offered as a general option.
     */
    CORPUS_1_0_COMPATIBILITY,
}

/** Everything the flattener needs that is not the record itself. */
data class FlattenOptions(
    val resolver: TypeResolver,
    val nfc: Nfc = PlatformNfc,
    val emptyContainers: EmptyContainerAuthorization = EmptyContainerAuthorization.REQUIRED,
)

/**
 * ROAX-CANON/1 section 3.3.
 *
 * ```
 * flatten(node, path):
 *   if node is a map and is empty:        emit (path, typeTag(path, "object"), -)
 *   if node is a map and is non-empty:    for each key k: flatten(node[k], path ‖ KEY(k))
 *   if node is an array and is empty:     emit (path, typeTag(path, "array"), -)
 *   if node is an array and is non-empty: for each index i: flatten(node[i], path ‖ INDEX(i))
 *   otherwise:                            emit (path, typeTag(path, jsonKind(node)), node)
 * ```
 *
 * Map iteration order is irrelevant, because leaves are sorted by encoded path in section 9. That
 * is deliberate: it removes the class of bug OpenAttestation inherits from JavaScript key
 * enumeration, and it is why a Go implementation works despite Go randomising map iteration.
 *
 * This produces the record's leaves **only**. The leaf set is the union of these and the reserved
 * leaves of section 11.2, and the union is formed before the sort - see [commit].
 */
fun flattenRecord(record: JsonValue, options: FlattenOptions): List<FlatLeaf> {
    val out = ArrayList<FlatLeaf>()
    flatten(record, emptyList(), options, out)
    return out
}

private fun flatten(
    node: JsonValue,
    segments: List<Segment>,
    options: FlattenOptions,
    out: MutableList<FlatLeaf>,
) {
    when (node) {
        is JsonObject -> {
            if (node.isEmpty) {
                out.add(FlatLeaf(segments, emptyContainerTag(segments, JsonKind.OBJECT, options), RoaxValue.EmptyObject))
                return
            }
            for (member in node.members) {
                // The guard checks the FIRST segment only, because every reserved path is a single
                // segment (section 11.2).
                if (segments.isEmpty()) Reserved.guardFirstSegment(member.key, options.nfc)
                flatten(member.value, segments + Segment.Key(member.key), options, out)
            }
        }

        is JsonArray -> {
            if (node.isEmpty) {
                out.add(FlatLeaf(segments, emptyContainerTag(segments, JsonKind.ARRAY, options), RoaxValue.EmptyArray))
                return
            }
            node.elements.forEachIndexed { i, element ->
                flatten(element, segments + Segment.Index(i.toLong()), options, out)
            }
        }

        else -> out.add(scalarLeaf(node, segments, options))
    }
}

private fun emptyContainerTag(
    segments: List<Segment>,
    kind: JsonKind,
    options: FlattenOptions,
): TypeTag {
    val fallback = if (kind == JsonKind.ARRAY) TypeTag.EMPTY_ARRAY else TypeTag.EMPTY_OBJECT
    val resolved = options.resolver.resolve(segments, kind)
    if (resolved != null) {
        if (resolved != fallback) {
            fail(
                Reason.TYPE_TAG_KIND_MISMATCH,
                "${displayPath(segments)} is an empty ${kind.label} but the map binds " +
                    "tag ${resolved.code} ${resolved.name} there",
            )
        }
        return resolved
    }
    return when (options.emptyContainers) {
        EmptyContainerAuthorization.REQUIRED -> failClosed(segments, kind)
        EmptyContainerAuthorization.CORPUS_1_0_COMPATIBILITY -> fallback
    }
}

private fun scalarLeaf(node: JsonValue, segments: List<Segment>, options: FlattenOptions): FlatLeaf {
    val kind = JsonKind.of(node)
    val tag = options.resolver.resolve(segments, kind) ?: failClosed(segments, kind)

    // The tag comes from the schema, not from the JSON literal's syntax (section 4). What is
    // checked here is only that the tag can CARRY the observed kind: a map that bound a string
    // path to INTEGER would otherwise silently reinterpret the value.
    val value: RoaxValue = when (node) {
        is JsonString -> when (tag) {
            TypeTag.STRING -> RoaxValue.Text(node.value)
            // The ruled FHIR `base64Binary` binding (grade Strong): BYTES over the DECODED octets,
            // with canonical RFC 4648 base64 as an input-admissibility condition. The octets are
            // what gets hashed; a spelling outside the pinned form is refused before it is decoded.
            TypeTag.BYTES -> RoaxValue.Bytes(Base64Strict.decode(node.value))
            else -> tagKindMismatch(segments, tag, kind)
        }

        is JsonNumber -> when (tag) {
            TypeTag.INTEGER -> RoaxValue.Integer(node.literal)
            TypeTag.DECIMAL -> RoaxValue.Decimal(node.literal)
            else -> tagKindMismatch(segments, tag, kind)
        }

        is JsonBoolean -> when (tag) {
            TypeTag.BOOL -> RoaxValue.Bool(node.value)
            else -> tagKindMismatch(segments, tag, kind)
        }

        is JsonNull -> when (tag) {
            TypeTag.NULL -> RoaxValue.Null
            else -> tagKindMismatch(segments, tag, kind)
        }

        else -> throw IllegalStateException("container reached the scalar path")
    }

    if (tag == TypeTag.BLOB_REF) {
        fail(
            Reason.BLOB_REF_NOT_DECLARED,
            "${displayPath(segments)} resolves to tag 8 BLOB_REF, which no version-1 profile " +
                "declares (specification section 6.5)",
        )
    }
    return FlatLeaf(segments, tag, value)
}

private fun failClosed(segments: List<Segment>, kind: JsonKind): Nothing = fail(
    Reason.TYPE_MAP_FAIL_CLOSED,
    "the selected type map has no binding for ${displayPath(segments)} at observed kind " +
        "'${kind.label}' (specification section 4.2, ruled decision D7)",
)

private fun tagKindMismatch(segments: List<Segment>, tag: TypeTag, kind: JsonKind): Nothing = fail(
    Reason.TYPE_TAG_KIND_MISMATCH,
    "the map binds ${displayPath(segments)} to tag ${tag.code} ${tag.name}, which cannot carry " +
        "an observed ${kind.label}",
)
